expand: Simplify expansion of derives

petrochenkov opened the pull request:
And make it more uniform with other macros.
This is done by merging placeholders for future derives' outputs into the derive container's output fragment early (addressing FIXMEs from https://github.com/rust-lang/rust/pull/63667).

Also, macros with names starting with `_` are no longer reported as unused, in accordance with the usual behavior of `unused` lints.

r? @matthewjasper or @mark-i-m

petrochenkov committed:
resolve: Mark macros starting with an underscore as used

petrochenkov committed:
expand: Simplify expansion of derives

And make it more uniform with other macros.
By merging placeholders for future derives' outputs into the derive container's output fragment early.

petrochenkov commented:
FWIW, the end goal here is attempting to turn `derive` into a regular attribute macro, but there's still quite a bit of steps until that:
- Expanding derives immediately with their container instead of enqueueing them for later expansion. Maybe I'll do that in this PR if it stays unmerged for a few days.
- Implementing left-to-right attribute expansion (cc @CAD97). Right now derives (perhaps from multiple derive containers) are always expanded after attributes, regardless of their order in source code, that's unnatural and is a source of some hacks. We should try to fix it, thankfully there is some [future proofing](https://github.com/rust-lang/rust/pull/54277) in place.

mark-i-m commented on src/librustc_resolve/build_reduced_graph.rs:
```diff
@@ -1061,8 +1051,14 @@ impl<'a, 'b> BuildReducedGraphVisitor<'a, 'b> {
         None
     }
 
+    fn insert_unused_macro(&mut self, ident: Ident, node_id: NodeId, span: Span) {
```
Perhaps add a comment here? Would the following be correct?:

```
Remember that the given macro is unused unless its name starts with `_`. Later, we will print `unused_macros` lints for all of them.
```

mark-i-m commented on src/librustc_resolve/build_reduced_graph.rs:
```diff
@@ -161,25 +161,15 @@ impl<'a> Resolver<'a> {
-        }
-
         let mut visitor = BuildReducedGraphVisitor { r: self, parent_scope };
         fragment.visit_with(&mut visitor);
```
Just to test my understanding, this visit now takes care of the derive placeholders, right?

petrochenkov replied:
Yes, the placeholders for derives are now inside the fragment, like placeholders for other kinds of macros.

mark-i-m reviewed (commented):
Overall, it looks good to me and makes sense, but I don't feel confident enough in my understanding of derive macros to give a r=me...

petrochenkov commented on src/libsyntax/ext/expand.rs:
```diff
@@ -72,6 +71,20 @@ macro_rules! ast_fragments {
+                match self {
+                    $($(AstFragment::$Kind(ast) => ast.extend(placeholders.iter().flat_map(|id| {
+                        #[cfg_attr(bootstrap, allow(unused_macros))]
+                        macro _repeating($flat_map_ast_elt) {}
```
To mention the `$flat_map_ast_elt` token somehow.
Otherwise we get `error: attempted to repeat an expression containing no syntax variables matched as repeating at this depth` because the outer `macro_rules! ast_fragments` doesn't know what to repeat here.
Any suggestions on how to name `_repeating` in a more clear way, or should I just add a comment?

matthewjasper commented:
@bors r+

bors commented:
:pushpin: Commit 1270140c124a93ffdada2f2422560ea498f0e7e4 has been approved by `matthewjasper`

<!-- @bors r=matthewjasper 1270140c124a93ffdada2f2422560ea498f0e7e4 -->
<!-- homu: {"type":"Approved","sha":"1270140c124a93ffdada2f2422560ea498f0e7e4","approver":"matthewjasper"} -->

bors commented:
:hourglass: Testing commit 1270140c124a93ffdada2f2422560ea498f0e7e4 with merge 9f86b1832c885b9d61a34d977eb8ea2fc499faf9...
<!-- homu: {"type":"BuildStarted","head_sha":"1270140c124a93ffdada2f2422560ea498f0e7e4","merge_sha":"9f86b1832c885b9d61a34d977eb8ea2fc499faf9"} -->

rust-highfive commented:
Your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=10725&view=logs&jobId=1e116cca-0a74-5ed2-d113-bc13ce542b6b) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/10725/logs/578), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/10725/logs/578)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain

```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

bors commented:
:broken_heart: Test failed - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=10725)
<!-- homu: {"type":"BuildFailed","builder_url":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=10725","builder_name":"checks-azure"} -->

pietroalbini commented:
@bors retry azure network error

Centril commented:
@bors rollup=never (possible cause of failure in two rollups)

petrochenkov commented:
Cannot reproduce locally, let's see what CI will say.
@bors r=matthewjasper

bors commented:
:pushpin: Commit 6ae414333eb1d869638ec9d219245cd7d2236873 has been approved by `matthewjasper`

<!-- @bors r=matthewjasper 6ae414333eb1d869638ec9d219245cd7d2236873 -->
<!-- homu: {"type":"Approved","sha":"6ae414333eb1d869638ec9d219245cd7d2236873","approver":"matthewjasper"} -->

bors commented:
:hourglass: Testing commit 6ae414333eb1d869638ec9d219245cd7d2236873 with merge 4b64ea300c52305bef9780c39e1b9d3a85e44b63...
<!-- homu: {"type":"BuildStarted","head_sha":"6ae414333eb1d869638ec9d219245cd7d2236873","merge_sha":"4b64ea300c52305bef9780c39e1b9d3a85e44b63"} -->

bors commented:
:broken_heart: Test failed - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=10768)
<!-- homu: {"type":"BuildFailed","builder_url":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=10768","builder_name":"checks-azure"} -->

rust-highfive commented:
The job `dist-x86_64-linux` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=10768&view=logs&jobId=4f922444-fdfe-5dcf-b824-02f86439ef14) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/10768/logs/900), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/10768/logs/900)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2019-10-15T06:36:38.9477629Z note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
2019-10-15T06:36:39.0512108Z error: Could not document `syntax`.
2019-10-15T06:36:39.0517653Z 
2019-10-15T06:36:39.0517779Z Caused by:
2019-10-15T06:36:39.0530499Z   process didn't exit successfully: `/checkout/obj/build/bootstrap/debug/rustdoc --edition=2018 --crate-type lib --crate-name syntax src/libsyntax/lib.rs --target x86_64-unknown-linux-gnu -o /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/doc --color always -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/release/deps --extern bitflags=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps/libbitflags-28fa941764cd0b5d.rmeta --extern lazy_static=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps/liblazy_static-25e195a008dc650e.rmeta --extern log=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps/liblog-6a39080e603a7540.rmeta --extern rustc_data_structures=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_data_structures-8fe38918b9a2a938.rmeta --extern errors=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_errors-f5bdf664f2735d32.rmeta --extern rustc_index=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_index-da81bc8f818c54ab.rmeta --extern rustc_lexer=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_lexer-96f153401da80175.rmeta --extern rustc_target=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_target-e6bb297dc731315e.rmeta --extern scoped_tls=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps/libscoped_tls-7c3a2b01c74539dd.rmeta --extern rustc_serialize=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps/libserialize-95bcadfb6805bd96.rmeta --extern smallvec=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps/libsmallvec-b7c252bfcc11690e.rmeta --extern syntax_pos=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps/libsyntax_pos-b59e470af0b21954.rmeta --document-private-items --passes strip-hidden` (exit code: 1)
2019-10-15T06:36:45.1756638Z [RUSTC-TIMING] syntax test:false 10.807
2019-10-15T06:36:45.1892085Z error: build failed
2019-10-15T06:36:45.1914488Z 
2019-10-15T06:36:45.1915587Z 
2019-10-15T06:36:45.1915587Z 
2019-10-15T06:36:45.1918195Z command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "doc" "-Zconfig-profile" "--target" "x86_64-unknown-linux-gnu" "-Zbinary-dep-depinfo" "-j" "2" "--release" "--locked" "--color" "always" "--features" "jemalloc" "--manifest-path" "/checkout/src/rustc/Cargo.toml" "--no-deps" "-p" "rustc" "-p" "syntax" "-p" "syntax_pos" "-p" "rustc_save_analysis" "-p" "rustc_plugin" "-p" "rustc_lint" "-p" "graphviz" "-p" "rustc_codegen_utils" "-p" "build_helper" "-p" "rustc_llvm" "-p" "rustc_metadata" "-p" "rustc_index" "-p" "serialize" "-p" "arena" "-p" "rustc_macros" "-p" "fmt_macros" "-p" "rustc_plugin_impl" "-p" "rustc_passes" "-p" "rustc_apfloat" "-p" "rustc_privacy" "-p" "rustc_mir" "-p" "rustc_lexer" "-p" "rustc_errors" "-p" "rustc_incremental" "-p" "syntax_ext" "-p" "rustc_resolve" "-p" "rustc_codegen_llvm" "-p" "rustc_target" "-p" "rustc_typeck" "-p" "rustc_interface" "-p" "rustc_data_structures" "-p" "rustc_codegen_ssa" "-p" "rustc_fs_util" "-p" "rustc_driver" "-p" "rustc_traits"
2019-10-15T06:36:45.1918820Z 
2019-10-15T06:36:45.1918865Z 
2019-10-15T06:36:45.1933358Z failed to run: /checkout/obj/build/bootstrap/debug/bootstrap dist --host x86_64-unknown-linux-gnu --target x86_64-unknown-linux-gnu
2019-10-15T06:36:45.1933496Z Build completed unsuccessfully in 1:31:35
2019-10-15T06:36:45.1933496Z Build completed unsuccessfully in 1:31:35
2019-10-15T06:36:45.1999935Z == clock drift check ==
2019-10-15T06:36:45.2019773Z   local time: Tue Oct 15 06:36:45 UTC 2019
2019-10-15T06:36:45.4473188Z   network time: Tue, 15 Oct 2019 06:36:45 GMT
2019-10-15T06:36:45.4473545Z == end clock drift check ==
2019-10-15T06:36:46.8921931Z ##[error]Bash exited with code '1'.
2019-10-15T06:36:46.8959790Z ##[section]Starting: Upload CPU usage statistics
2019-10-15T06:36:46.8968136Z ==============================================================================
2019-10-15T06:36:46.8968255Z Task         : Bash
2019-10-15T06:36:46.8968347Z Description  : Run a Bash script on macOS, Linux, or Windows
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

petrochenkov commented:
`rustdoc` doesn't like `macro` items inside functions, which is weird because it was supposed to be fixed by https://github.com/rust-lang/rust/pull/63624, I'll check further.

petrochenkov committed:
privacy: Avoid one more `unwrap` causing an ICE in rustdoc

The issue is rustdoc-specific because its root cause if the `everybody_loops` pass makes some def-ids to not have local hir-ids

petrochenkov commented:
Removed one more `unwrap` from the function that https://github.com/rust-lang/rust/pull/63624 previously fixed.
@bors r=matthewjasper

bors commented:
:pushpin: Commit 487e3c1ca90d0bdc73ff479b4fda24d888b3cf54 has been approved by `matthewjasper`

<!-- @bors r=matthewjasper 487e3c1ca90d0bdc73ff479b4fda24d888b3cf54 -->
<!-- homu: {"type":"Approved","sha":"487e3c1ca90d0bdc73ff479b4fda24d888b3cf54","approver":"matthewjasper"} -->

bors commented:
:umbrella: The latest upstream changes (presumably #65495) made this pull request unmergeable. Please resolve the merge conflicts.

petrochenkov committed:
Fix rebase

petrochenkov commented:
@bors r=matthewjasper

bors commented:
:pushpin: Commit 7f89f04b41d4d1de85df2fd3ede61e4fc97d9e69 has been approved by `matthewjasper`

<!-- @bors r=matthewjasper 7f89f04b41d4d1de85df2fd3ede61e4fc97d9e69 -->
<!-- homu: {"type":"Approved","sha":"7f89f04b41d4d1de85df2fd3ede61e4fc97d9e69","approver":"matthewjasper"} -->

petrochenkov commented:
@bors rollup=maybe
